### PR TITLE
API improvements (rescale, params.init, params.scope, sgd)

### PR DESCRIPTION
**Pull request by ry**

### CODE REVIEW

File: src/tensor.ts [hunk 1]
```diff
@@ -550,6 +552,69 @@ export class Tensor implements types.BasicTensor {
...
+  /** To rescale a tensor from inScale to outScale.
+   * This is often used rescale images from [0, 255] integer
+   * range to [-1, 1] range. This function will automatically cast tensors to
+   * float32.
+   *
+   *    import { T } from "propel";
+   *    let pretendImage = T([0, 127, 255], "int32");
+   *    pretendImage.rescale([0, 255], [-1, 1]);
+   */
+  rescale(inScale: number[], outScale: number[]): Tensor {
```
Comment (piscisaureus) [inline src/tensor.ts]: Use type `[number, number]`

File: src/tensor.ts [hunk 2]
```diff
@@ -550,6 +552,69 @@ export class Tensor implements types.BasicTensor {
...
+   * If users are calling this function multiple times, they should use
+   * params.scope("myLayerName") to not conflict with other params.
+   *
+   *    import { params, zeros } from "propel";
+   *    params = params();
+   *    inputs = zeros([2, 5]);
+   *    outputs = inputs.linear(params.scope("L1"), 10);
+   *    params.has("L1/weights") && params.has("L1/bias");
+   */
+  linear(params: Params, outDim: number, opts: LinearOpts = {}): Tensor {
```
Comment (piscisaureus) [inline src/tensor.ts]: Note that you can place a default value for all opts here:

```ts
linear(params: Params, outDim: number, { bias = true, scale = 0.01 }: LinearOpts = {}): Tensor {
```

File: src/api.ts [hunk 3]
```diff
@@ -157,84 +163,32 @@ export function ones(shape: types.Shape,
...
-    // Initialize.
-    let t = zeros(shape);
-    if (device && device !== "CPU:0") {
-      t = new Tensor(bo.copyToDevice(t.basic, device));
+    for (const name of Object.keys(grads)) {
+      const g = grads[name];
+      const p = params.get(name);
+      // p -= g * lr
+      p.assign(p.sub(g.mul(opts.lr)));
+      keep(p);
```
Comment (piscisaureus) [inline src/api.ts]: Is keep() necessary here. I would expect the tensor to be kept implicitly, since it existed before the call to gc().

Comment (ry) [reply src/api.ts]: Good point - removed.

File: src/tensor.ts (hunk 2 above)
Comment (ry) [reply src/tensor.ts]: done

File: src/tensor.ts (hunk 1 above)
Comment (ry) [reply src/tensor.ts]: done

**Pull request merged**